lib/AdapterLdap.php

```diff
@@ -255,58 +287,87 @@ public function getFacilityByEntityId($spEntityId)
 
     public function getEntitylessAttribute($attrName)
     {
-        throw new BadMethodCallException('NotImplementedException');
-        // TODO: Implement getEntitylessAttribute() method.
+        return $this->fallbackAdapter->getEntitylessAttribute($attrName);
     }
 
     public function getVoAttributes($vo, $attrNames)
     {
-        throw new BadMethodCallException('NotImplementedException');
-        // TODO: Implement getVoAttribute() method.
+        return $this->fallbackAdapter->getVoAttributes($vo, $attrNames);
+    }
+
+    public function getVoAttributesValues($vo, $attributes)
+    {
+        $attrTypeMap = AttributeUtils::createLdapAttrNameTypeMap($attributes);
+
+        $perunAttrs = $this->connector->searchForEntities(
+            $this->ldapBase,
+            '(&(objectClass=perunVO)(perunVoId=' . $vo->getId() . '))',
+            array_keys($attrTypeMap)
+        );
+
+        $attributesValues = [];
+
+        foreach (array_keys($attrTypeMap) as $attrName) {
+            $attributesValues[$attrTypeMap[$attrName][self::INTERNAL_ATTR_NAME]] =
+                $this->setAttrValue($attrTypeMap, $perunAttrs[0], $attrName);
+        }
+
+        return $attributesValues;
     }
 
     public function getFacilityAttribute($facility, $attrName)
     {
-        throw new BadMethodCallException('NotImplementedException');
-        // TODO: Implement getFacilityAttribute() method.
+        return $this->fallbackAdapter->getFacilityAttribute($facility, $attrName);
```

Comment:
just a question out of curiosity, why are not some methods implemented and use fallback adapter? As far as I know, most of the attributes are already present in the LDAP and therefore we can implement these methods.

Reply:
AFAIK we're not able to get a whole attribute as in RPC (I mean with all its attributes as friendlyName, namespace, type, createdAt etc.). That's why we don't have just "getXAttributes" methods, but "getXAttributesValues", too. All "getXAttributesValues" methods should be implemented, "getXAttributes" methods fallback to RPC where they try to get a whole attribute.